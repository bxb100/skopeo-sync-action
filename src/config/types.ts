import { z } from 'zod'

export const AuthSchema = z.object({
  username: z.string(),
  password: z.string()
})

export const AuthMapSchema = z.record(z.string(), AuthSchema)

const _DestSchema = z
  .string()
  .or(z.string().array())
  .transform(v => {
    if (typeof v === 'string') {
      return [v]
    }
    return v
  })

export const ImageSyncSemverSchema = z.object({
  semver: z.string().or(z.number().transform(String)),
  dest: _DestSchema
})

export const ImageSyncRegexSchema = z.object({
  regex: z.string(),
  dest: _DestSchema
})

export const ImageSyncValSchema = _DestSchema
  .or(ImageSyncSemverSchema)
  .or(ImageSyncRegexSchema)

export const ImageSyncMapSchema = z.record(z.string(), ImageSyncValSchema)

export enum SyncType {
  All,
  Tag,
  Digest,
  Regex,
  Semver
}

export type ImageSyncInfo = {
  type: SyncType
  image: string
  tag?: string[]
  regex?: RegExp
  digest?: string
  semver?: string
}

export const SkopeoListTagsResSchema = z.object({
  Repository: z.string(),
  Tags: z.string().array()
})

export class Sync {
  source_image: string
  dest_image: string
  private readonly ori_s: string
  private readonly ori_d: string

  constructor(source_image: string, dest_image: string) {
    this.ori_s = source_image
    this.ori_d = dest_image
    this.source_image = `docker://${source_image}`
    this.dest_image = `docker://${dest_image}`
  }

  fmt(): string {
    const d = this.ori_d.split('/')[0]
    return `sync ${this.ori_s} to ${d}`
  }
}
