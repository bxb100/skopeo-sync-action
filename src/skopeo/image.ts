import {
  ImageSyncInfo,
  ImageSyncValSchema,
  SkopeoListTagsResSchema,
  Sync,
  SyncType
} from './types'
import { exec } from '@actions/exec'
import { z } from 'zod'
import * as core from '@actions/core'
import { satisfies } from 'semver'

/**
 * https://docs.docker.com/reference/cli/docker/image/pull/
 *
 * `NAME[:TAG|@DIGEST]`
 *
 * Addition support `NAME:/regex/`
 *
 * @param source_image eg. `nginx:latest`, `ghcr.io/advplyr/audiobookshelf:latest`
 */
export function sync_type(source_image: string): ImageSyncInfo {
  if (source_image.includes('@')) {
    const [image, digest] = source_image.split('@')
    return {
      type: SyncType.Digest,
      image,
      digest
    }
  }
  const [image, tag] = source_image.split(':')
  if (tag) {
    if (tag.startsWith('/') && tag.endsWith('/')) {
      return {
        type: SyncType.Regex,
        image,
        regex: new RegExp(tag)
      }
    }
    return {
      type: SyncType.Tag,
      image,
      tag: tag.split(',')
    }
  }
  return {
    type: SyncType.All,
    image
  }
}

export async function pair(
  k: string,
  v: z.infer<typeof ImageSyncValSchema>
): Promise<Sync[]> {
  let sync_info: ImageSyncInfo
  let dest_image: string[]
  if ('semver' in v) {
    sync_info = {
      image: k,
      type: SyncType.Semver,
      semver: v.semver
    }
    dest_image = v.dest
  } else if ('regex' in v) {
    sync_info = {
      image: k,
      type: SyncType.Regex,
      regex: new RegExp(v.regex)
    }
    dest_image = v.dest
  } else {
    sync_info = sync_type(k)
    dest_image = v
  }

  console.log(sync_info)

  const source_images = await gen_need_sync_source_images(sync_info, dest_image)
  core.info(`source images: ${JSON.stringify(source_images)}`)
  if (source_images === undefined) {
    throw new Error(`Image ${k} not found`)
  }
  return source_images
}

export async function copy(syncs: Sync[], skip_error: boolean): Promise<void> {
  core.summary.addHeading('Sync Summary:', 2)
  for (let i = 0; i < syncs.length; i++) {
    const sync = syncs[i]
    const err_line: string[] = []
    const out_line: string[] = []
    // https://github.com/containers/skopeo/blob/main/docs/skopeo-copy.1.md
    const exit_code = await exec(
      'skopeo',
      // --multi-arch not support under 1.8.0
      // https://github.com/containers/skopeo/commit/4ef35a385af074c979c9f8c4e2e37c38b0963c3a
      // https://github.com/actions/runner-images ubuntu-latest(ubuntu-22.04) skopeo-1.4.1
      ['copy', '--all', sync.source_image, sync.dest_image],
      {
        listeners: {
          errline: (data: string) => err_line.push(data),
          stdline: (data: string) => out_line.push(data)
        }
      }
    )
    core.summary.addRaw(
      `process [${i + 1}/${syncs.length}]: ${sync.fmt()}`,
      true
    )
    if (err_line.length > 0) {
      core.summary.addDetails(':x:', '\n'.repeat(2) + err_line.join('\n'))
    } else if (out_line.length > 0) {
      core.summary.addDetails(
        ':white_check_mark:',
        '\n'.repeat(2) + out_line.join('\n')
      )
    }
    await core.summary.write()
    if (!skip_error && exit_code != 0) {
      throw new Error(
        `sync ${sync.source_image} to ${sync.dest_image} failed: ${err_line.join('\n')}`
      )
    }
  }
}

/**
 * https://github.com/containers/skopeo/blob/main/docs/skopeo-list-tags.1.md
 *
 * @param source_image_name eg. `docker.io/fedora`, `nginx`
 */
export async function list_tag(
  source_image_name: string
): Promise<z.infer<typeof SkopeoListTagsResSchema>> {
  let res = ''
  let err = ''
  const exit_code = await exec(
    'skopeo',
    ['list-tags', `docker://${source_image_name}`],
    {
      listeners: {
        stdout: (data: Buffer) => {
          res += data.toString()
        },
        stderr: (data: Buffer) => {
          err += data.toString()
        }
      }
    }
  )

  if (exit_code != 0) {
    throw new Error(`skopeo list-tags: ${err}`)
  }

  return SkopeoListTagsResSchema.parse(JSON.parse(res))
}

export async function gen_need_sync_source_images(
  sync_info: ImageSyncInfo,
  dest_images: string[]
): Promise<Sync[] | undefined> {
  if (sync_info.type === SyncType.Tag) {
    return sync_info.tag
      ?.map(t => `${sync_info.image}:${t}`)
      .flatMap(s => dest_images.map(d => new Sync(s, d)))
  }

  if (sync_info.type === SyncType.Digest) {
    return dest_images.map(
      d => new Sync(`${sync_info.image}@${sync_info.digest}`, d)
    )
  }

  const tag = await list_tag(sync_info.image)

  return dest_images.flatMap(dest => {
    switch (sync_info.type) {
      case SyncType.All:
        return tag.Tags.map(t => new Sync(sync_info.image, dest, t))
      case SyncType.Semver:
        return tag.Tags.filter(t =>
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          satisfies(t, sync_info.semver!)
        ).map(t => new Sync(sync_info.image, dest, t))
      case SyncType.Regex:
        return tag.Tags.filter(t => sync_info.regex?.test(t)).map(
          t => new Sync(sync_info.image, dest, t)
        )
    }
    throw new Error(`Unknown error ${sync_info.type}`)
  })
}
