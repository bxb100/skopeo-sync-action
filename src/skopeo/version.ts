import { exec } from '@actions/exec'

export async function print_skopeo_version(): Promise<void> {
  await exec('skopeo', ['--version'])
}
