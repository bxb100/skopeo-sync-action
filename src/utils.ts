import * as fs from 'node:fs'
import { ZodType } from 'zod'
import { parseDocument } from 'yaml'
import * as process from 'node:process'

export function parse_yaml<T>(file_path: string, schema: ZodType<T>): T {
  const file = fs.readFileSync(file_path, 'utf8')
  const doc = parseDocument(file)
  return schema.parse(doc.toJS())
}

/**
 * Simply inject env to template string, support `${ENV_NAME}`, `$ENV_NAME`
 *
 * @param template string
 */
export function inject_env(template: string): string {
  const envs = process.env
  const env_map = new Map<string, string | undefined>()
  for (const k in envs) {
    env_map.set(k.toLowerCase(), envs[k])
  }

  const tokens = lexer(template)
  const replace_map = new Map<string, string>()
  tokens.forEach(token => {
    const v = env_map.get(token.k.toLowerCase())
    if (v) {
      replace_map.set(template.substring(token.start, token.end + 1), v)
    }
  })
  let res = template
  for (const [k, v] of replace_map) {
    res = res.replace(k, v)
  }
  return res
}

export type Token = {
  start: number
  end: number
  pre_len: number
  k: string
}

export function lexer(str: string, strict = false): Token[] {
  const res: Token[] = []
  const keyword = '$'
  const bracket_pre = ['[', '(', '{']
  const bracket_pos = [']', ')', '}']
  const chars = Array.from(str)

  let c
  for (let i = 0; i < chars.length; i++) {
    c = chars[i]
    if (c === keyword) {
      i++
      let around_len = 0
      let pre: string | undefined
      let pos: string | undefined
      for (let j = i; j < i + 2; j++) {
        if (bracket_pre.includes(chars[j])) {
          if (strict && pre === undefined) {
            pre = chars[j]
            pos = bracket_pos[bracket_pre.indexOf(pre)]
          } else if (strict && pre) {
            if (chars[j] != pre) {
              throw new Error(`Invalid bracket, pre: ${pre}, cur: ${chars[j]}`)
            }
          }
          around_len++
        } else break
      }
      const start = i + around_len
      let j = start
      while (chars.length > j && /[a-zA-Z0-9_-]/.test(chars[j])) {
        j++
      }
      if (j === start) {
        // no valid alphanumeric ident
        continue
      }
      for (let k = j; k < j + around_len; k++) {
        if (k >= chars.length || !bracket_pos.includes(chars[k])) {
          throw new Error('Not valid sentence')
        }
        if (strict && pos && pos !== chars[k]) {
          throw new Error(
            `Invalid correspond bracket, except: ${pos}, cur: ${chars[k]}`
          )
        }
      }
      res.push({
        start: start - around_len - 1,
        end: j + around_len - 1,
        pre_len: around_len,
        k: str.substring(start, j)
      })
      // skip
      i = j - 1
    }
  }

  return res
}
