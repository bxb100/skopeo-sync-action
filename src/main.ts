import * as core from '@actions/core'
import { parse_yaml } from './utils'
import { AuthMapSchema, ImageSyncMapSchema } from './config/types'
import { login } from './config/auth'
import { copy } from './config/image'
import { print_skopeo_version } from './config/version'

export async function run(): Promise<void> {
  try {
    const auth_file = core.getInput('auth_file')
    const images_file = core.getInput('images_file')
    const skip_error = core.getBooleanInput('skip_error')

    await print_skopeo_version()

    const auths = parse_yaml(auth_file, AuthMapSchema)
    for (const registry in auths) {
      const login_status = await login(registry, auths[registry])
      core.info(`Login ${registry} ${login_status ? 'Success' : 'Failed'}`)
    }

    const image_sync_map = parse_yaml(images_file, ImageSyncMapSchema)
    for (const source_img in image_sync_map) {
      const dest_images = image_sync_map[source_img] as string[]
      await copy(source_img, dest_images, skip_error)
    }
  } catch (e) {
    core.setFailed(e as Error)
  }
}
