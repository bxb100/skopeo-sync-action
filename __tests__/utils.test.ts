import { inject_env, parse_yaml } from '../src/utils'
import { expect } from '@jest/globals'
import { AuthMapSchema, ImageSyncMapSchema } from '../src/skopeo/types'

describe('utils', () => {
  it('parse auth yaml', () => {
    const record = parse_yaml('./test/auth.yml', AuthMapSchema)
    expect(record).toMatchSnapshot()
  })

  it('parse image yaml', () => {
    const record = parse_yaml('./test/images.yml', ImageSyncMapSchema)
    console.log(record)
    expect(record).toMatchSnapshot()
  })

  it('inject env', () => {
    process.env.username = 'cs1'
    process.env.password = 'cs2'
    expect(inject_env('  username: ${username}')).toEqual('  username: cs1')
    expect(
      inject_env('  username: ${username} password: ${password} ')
    ).toEqual('  username: cs1 password: cs2 ')
    expect(inject_env('  username: ${px}')).toEqual('  username: ${px}')
    expect(inject_env('  username: $username')).toEqual('  username: cs1')
    expect(inject_env('  username: $px')).toEqual('  username: $px')
  })
})
