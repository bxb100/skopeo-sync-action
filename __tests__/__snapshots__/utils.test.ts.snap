// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils parse auth yaml 1`] = `
{
  "ghcr.io": {
    "password": "\${GITHUB_P}",
    "username": "\${GITHUB_U}",
  },
  "registry.cn-hangzhou.aliyuncs.com": {
    "password": "\${ALI_P}",
    "username": "\${ALI_U}",
  },
  "swr.cn-north-4.myhuaweicloud.com": {
    "password": "\${HUAWEI_P}",
    "username": "\${HUAWEI_U}",
  },
}
`;

exports[`utils parse image yaml 1`] = `
{
  "alpine": {
    "dest": [
      "swr.cn-north-4.myhuaweicloud.com/bxb100/alpine",
    ],
    "semver": ">= 3.20",
  },
  "ghcr.io/advplyr/audiobookshelf:latest": [
    "registry.cn-hangzhou.aliyuncs.com/bxb100/audiobookshelf:latest",
    "swr.cn-north-4.myhuaweicloud.com/bxb100/advplyr/audiobookshelf",
  ],
  "nginx": {
    "dest": [
      "swr.cn-north-4.myhuaweicloud.com/bxb100/nginx",
    ],
    "regex": "^1\\.13\\.[12]-alpine-perl$",
  },
}
`;
